File: master-server/raft/raft_fsm_candidate.go
package raft

import (
	"fmt"

	"master-server/raft/logger"
	"master-server/raft/proto"
)

func (r *raftFsm) becomeCandidate() {
	if r.state == stateLeader {
		panic(AppPanicError(fmt.Sprintf("[raft->becomeCandidate][%v] invalid transition [leader -> candidate].", r.id)))
	}

	r.step = stepCandidate
	r.reset(r.term+1, 0, false)
	r.tick = r.tickElection
	r.vote = r.config.NodeID
	r.state = stateCandidate

	if logger.IsEnableInfo() {
		logger.Info("raft[%v] became candidate at term %d.", r.id, r.term)
	}
}

func stepCandidate(r *raftFsm, m *proto.Message) {
	switch m.Type {
	case proto.LocalMsgProp:
		if logger.IsEnableDebug() {
			logger.Debug("raft[%v] no leader at term %d; dropping proposal.", r.id, r.term)
		}
		proto.ReturnMessage(m)
		return

	case proto.ReqMsgAppend:
		r.becomeFollower(r.term, m.From)
		r.handleAppendEntries(m)
		proto.ReturnMessage(m)
		return

	case proto.ReqMsgHeartBeat:
		r.becomeFollower(r.term, m.From)
		return

	case proto.ReqMsgElectAck:
		r.becomeFollower(r.term, m.From)
		nmsg := proto.GetMessage()
		nmsg.Type = proto.RespMsgElectAck
		nmsg.To = m.From
		r.send(nmsg)
		proto.ReturnMessage(m)
		return

	case proto.ReqMsgVote:
		if logger.IsEnableInfo() {
			logger.Info("raft[%v] [logterm: %d, index: %d, vote: %v] rejected vote from %v [logterm: %d, index: %d] at term %d.", r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.vote, m.From, m.LogTerm, m.Index, r.term)
		}
		nmsg := proto.GetMessage()
		nmsg.Type = proto.RespMsgVote
		nmsg.To = m.From
		nmsg.Reject = true
		r.send(nmsg)
		proto.ReturnMessage(m)
		return

	case proto.RespMsgVote:
		gr := r.poll(m.From, !m.Reject)
		if logger.IsEnableInfo() {
			logger.Info("raft[%v] [q:%d] has received %d votes and %d vote rejections.", r.id, r.quorum(), gr, len(r.votes)-gr)
		}
		switch r.quorum() {
		case gr:
			if r.config.LeaseCheck {
				r.becomeElectionAck()
			} else {
				r.becomeLeader()
				r.bcastAppend()
			}
		case len(r.votes) - gr:
			r.becomeFollower(r.term, NoLeader)
		}
	}
}

func (r *raftFsm) campaign(force bool) {
	r.becomeCandidate()
	if r.quorum() == r.poll(r.config.NodeID, true) {
		if r.config.LeaseCheck {
			r.becomeElectionAck()
		} else {
			r.becomeLeader()
		}
		return
	}

	for id := range r.replicas {
		if id == r.config.NodeID {
			continue
		}
		li, lt := r.raftLog.lastIndexAndTerm()
		if logger.IsEnableDebug() {
			logger.Debug("raft[%v] campaign: [logterm: %d, index: %d] sent vote request to %v at term %d.", r.id, lt, li, id, r.term)
		}

		m := proto.GetMessage()
		m.To = id
		m.Type = proto.ReqMsgVote
		m.ForceVote = force
		m.Index = li
		m.LogTerm = lt
		r.send(m)
	}
}

func (r *raftFsm) poll(id uint64, v bool) (granted int) {
	if logger.IsEnableDebug() {
		if v {
			logger.Debug("raft[%v] received vote from %v at term %d.", r.id, id, r.term)
		} else {
			logger.Debug("raft[%v] received vote rejection from %v at term %d.", r.id, id, r.term)
		}
	}
	if _, ok := r.votes[id]; !ok {
		r.votes[id] = v
	}
	for _, vv := range r.votes {
		if vv {
			granted++
		}
	}
	return granted
}

File: data-server/src/raft/src/impl/raft_fsm_candidate.cpp
#include "raft_fsm.h"

#include <functional>
#include "logger.h"
#include "raft_exception.h"

namespace sharkstore {
namespace raft {
namespace impl {

void RaftFsm::becomeCandidate() {
    if (state_ == FsmState::kLeader) {
        throw RaftException(
            "[raft->becomeCandidate] invalid transition [leader -> candidate]");
    }

    step_func_ = std::bind(&RaftFsm::stepCandidate, this, std::placeholders::_1);
    reset(term_ + 1, false);
    tick_func_ = std::bind(&RaftFsm::tickElection, this);
    vote_for_ = node_id_;
    state_ = FsmState::kCandidate;

    LOG_INFO("raft[%llu] become candidate at term %llu", id_, term_);
}

void RaftFsm::becomePreCandidate() {
    if (state_ == FsmState::kLeader) {
        throw RaftException("invalid transition[leader -> pre-candidate]");
    }

    step_func_ = std::bind(&RaftFsm::stepCandidate, this, std::placeholders::_1);
    tick_func_ = std::bind(&RaftFsm::tickElection, this);
    state_ = FsmState::kPreCandidate;
    votes_.clear();

    LOG_INFO("raft[%llu] become pre-candidate at term %llu", id_, term_);
}

void RaftFsm::stepCandidate(MessagePtr& msg) {
    switch (msg->type()) {
        case pb::LOCAL_MSG_PROP:
            LOG_DEBUG("raft[%llu] no leader at term %llu; dropping proposal.", id_,
                      term_);
            return;

        case pb::APPEND_ENTRIES_REQUEST:
            becomeFollower(term_, msg->from());
            handleAppendEntries(msg);
            return;

        case pb::HEARTBEAT_REQUEST:
            becomeFollower(term_, msg->from());
            return;

        case pb::VOTE_RESPONSE:
        case pb::PRE_VOTE_RESPONSE: {
            bool pre = false;
            if (msg->type() == pb::PRE_VOTE_RESPONSE) {
                if (state_ != FsmState::kPreCandidate) return;
                pre = true;
            } else {
                if (state_ != FsmState::kCandidate) return;
            }

            int grants = poll(pre, msg->from(), !msg->reject());
            int rejects = votes_.size() - grants;

            LOG_INFO("raft[%llu] %s [q:%d] has received %d votes and %d vote "
                     "rejections",
                     id_, (pre ? "pre-campaign" : "campaign"), quorum(), grants, rejects);

            if (grants == quorum()) {  // 收到大多数投票
                if (pre) {
                    campaign(false);
                } else {
                    becomeLeader();
                    bcastAppend();
                }
            } else if (rejects == quorum()) {  // 大多数都拒绝了
                becomeFollower(term_, 0);
            }
            return;
        }

        default:
            return;
    }
}

void RaftFsm::campaign(bool pre) {
    if (pre) {
        becomePreCandidate();
    } else {
        becomeCandidate();
    }

    // 可能只有一个成员
    if (poll(pre, node_id_, true) == quorum()) {
        if (pre) {
            campaign(false);
        } else {
            becomeLeader();
        }
        return;
    }

    // 发送投票请求
    uint64_t li = 0, lt = 0;
    raft_log_->lastIndexAndTerm(&li, &lt);
    for (const auto& r : replicas_) {
        if (r.first == node_id_) continue;

        LOG_INFO("raft[%llu] %s: [logterm: %llu, index: %llu] sent vote "
                 "request to %llu at term %llu.",
                 id_, (pre ? "pre-campaign" : "campaign"), lt, li, r.first, term_);

        MessagePtr msg(new pb::Message);
        msg->set_type(pre ? pb::PRE_VOTE_REQUEST : pb::VOTE_REQUEST);
        // PreVote请求的term是当前的term+1，因为preCandidate状态不增加term
        msg->set_term(pre ? (term_ + 1) : term_);
        msg->set_to(r.first);
        msg->set_log_index(li);
        msg->set_log_term(lt);
        send(msg);
    }
}

int RaftFsm::poll(bool pre, uint64_t node_id, bool vote) {
    LOG_INFO("raft[%llu] received %s %s from %llu at term %llu", id_,
             (pre ? "pre-vote" : "vote"), (vote ? "granted" : "rejected"), node_id,
             term_);

    votes_.emplace(node_id, vote);

    int grants = 0;
    for (const auto& kv : votes_) {
        if (kv.second) ++grants;
    }
    return grants;
}

} /* namespace impl */
} /* namespace raft */
} /* namespace sharkstore */
